src/main/java/blockchains/iaas/uni/stuttgart/de/restapi/Controllers/PluginManagerController.java

```diff
@@ -0,0 +1,128 @@
+import org.glassfish.jersey.media.multipart.FormDataContentDisposition;
+import org.glassfish.jersey.media.multipart.FormDataParam;
+import org.pf4j.PluginWrapper;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.ws.rs.*;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.io.*;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+import javax.ws.rs.core.UriInfo;
+
+@Path("plugin-manager")
+public class PluginManagerController {
+    private static final Logger log = LoggerFactory.getLogger(PluginManagerController.class);
+
+    @Context
+    protected UriInfo uriInfo;
+
+    @POST
+    @Path("/upload-plugin")
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public Response uploadJar(@FormDataParam("file") InputStream uploadedInputStream,
+                              @FormDataParam("file") FormDataContentDisposition fileDetails) {
+
+        BlockchainPluginManager blockchainPluginManager = BlockchainPluginManager.getInstance();
+        String fileName = fileDetails.getFileName();
+        log.info("Received file {}", fileName);
+        String uploadedFileLocation = blockchainPluginManager.getPluginPath() + "/" + fileDetails.getFileName();
+        java.nio.file.Path filePath = Paths.get(uploadedFileLocation);
+        if (Files.exists(filePath)) {
+            log.error("Received file {} already exists in plugins directory.", fileName);
+            return Response.status(Response.Status.BAD_REQUEST).entity("File already exists with same name.").build();
+        }
+        writeToFile(uploadedInputStream, uploadedFileLocation);
+
+        blockchainPluginManager.loadJar(filePath);
+
+        return Response.ok().build();
+    }
+
+    @POST
+    @Path("/enable-plugin")
+    public Response enablePlugin() {
+        String pluginId = this.uriInfo.getQueryParameters().getFirst("plugin-id");
+        BlockchainPluginManager.getInstance().enablePlugin(pluginId);
+        return Response.ok().build();
+    }
+
+    @POST
+    @Path("/start-plugin")
+    public Response startPlugin() {
+        String pluginId = this.uriInfo.getQueryParameters().getFirst("plugin-id");
+        BlockchainPluginManager.getInstance().startPlugin(pluginId);
+        return Response.ok().build();
+    }
+
+    @POST
+    @Path("/disable-plugin")
+    public Response disablePlugin() {
+        String pluginId = this.uriInfo.getQueryParameters().getFirst("plugin-id");
+        BlockchainPluginManager.getInstance().disablePlugin(pluginId);
+        return Response.ok().build();
+    }
+
+    @POST
+    @Path("/unload-plugin")
+    public Response unloadPlugin() {
+        String pluginId = this.uriInfo.getQueryParameters().getFirst("plugin-id");
+        BlockchainPluginManager.getInstance().unloadPlugin(pluginId);
+        return Response.ok().build();
+    }
+
+    @DELETE
+    public Response deletePlugin() {
```

Comment:
What is the difference between deleting a plugin and unloading it?

Reply:
`delete` unloads a plugin and deletes the jar from the plugins directory. So, user has to upload a jar again. 
 
`unloading` means removing the plugin from the running application instance but, not deleting the jar. So, after application restart, the same jar will be loaded again and user can start the plugin.

Reply:
OK